src/metabase/driver/postgres.clj

```diff
@@ -209,6 +213,58 @@
 
 (def ^:private ^:dynamic *enum-types* nil)
 
+(defn- get-tables-sql
+  [schemas table-names]
+  ;; Ref: https://github.com/davecramer/pgjdbc/blob/a714bfd/pgjdbc/src/main/java/org/postgresql/jdbc/PgDatabaseMetaData.java#L1272
+  (sql/format
+   (cond-> {:select    [[:n.nspname :schema]
+                        [:c.relname :name]
+                        [[:case-expr :c.relkind
+                          [:inline "r"] [:inline "TABLE"]
+                          [:inline "p"] [:inline "PARTITIONED TABLE"]
+                          [:inline "v"] [:inline "VIEW"]
+                          [:inline "f"] [:inline "FOREIGN TABLE"]
+                          [:inline "m"] [:inline "MATERIALIZED VIEW"]
+                          :else nil]
+                         :type]
+                        [:d.description :description]]
+            :from      [[:pg_catalog.pg_namespace :n]
+                        [:pg_catalog.pg_class :c]]
+            :left-join [[:pg_catalog.pg_description :d] [:and [:= :c.oid :d.objoid] [:= :d.objsubid [:inline 0]]]
+                        [:pg_catalog.pg_class :dc]      [:and [:= :d.classoid :dc.oid] [:= :dc.relname [:inline "pg_class"]]]
+                        [:pg_catalog.pg_namespace :dn]  [:and [:= :dn.oid :dc.relnamespace] [:= :dn.nspname [:inline "pg_catalog"]]]]
+            :where     [:and [:= :c.relnamespace :n.oid]
+                        ;; filter out system tables
+                        [(keyword "!~") :n.nspname "^pg_"] [:<> :n.nspname "information_schema"]
+                        ;; only get tables of type: TABLE, PARTITIONED TABLE, VIEW, FOREIGN TABLE, MATERIALIZED VIEW
+                        [:raw "c.relkind in ('r', 'p', 'v', 'f', 'm')"]]
+            :order-by  [:type :schema :name]}
+     (seq schemas)
+     (sql.helpers/where [:in :n.nspname schemas])
+
+     (seq table-names)
+     (sql.helpers/where [:in :c.relname table-names]))
+   {:dialect :ansi}))
+
+(defn- get-tables
+  ;; have it as its own method for ease of testing
+  [database schemas tables]
+  (sql-jdbc.execute/reducible-query database (get-tables-sql schemas tables)))
+
+(defmethod driver/describe-database :postgres
+  [driver database]
+  {:tables
+   (sql-jdbc.execute/do-with-connection-with-options
```

Comment:
As Cam suggested, in the future it would be great to have a new multimethod that returned a reducible of results so we don't have keep all the tables in memory. While we're writing this code, maybe we should make that change as easy as possible.

I don't think we need to create the new multimethod yet, but we can prepare to switch in the future. I'd suggest a simple refactor here, creating a new function that returns a reducible collection of table metadata, and have `driver/describe-database` use it.

e.g.

```clojure
(defmethod driver/describe-database :postgres
  [driver database]
  ;; TODO: convert describe-syncable-tables into its own multimethod
  {:tables (into #{} (describe-syncable-tables driver database))
```

Then when we come to create the new multimethod, we can simply replace `(defn describe-syncable-tables ..)` with `(defmulti describe-syncable-tables ...)` or whatever we want to call it

Reply:
good idea 👍 [a84ee86](https://github.com/metabase/metabase/pull/39670/commits/a84ee8604d75e92f1a180d429114cbe61ec6ff1d)